231 refactor coclustering iv by removing management of antecedent

carinehue opened the pull request:
Implémentation proto supprimant l'utilisation de l'antecedent
Correction de bug mémoire et d'un bug de calcul de cout
Nouvelle version de LearningTest suite a ces corrections

A merger après la PR https://github.com/KhiopsML/khiops/pull/459

carinehue committed:
Remove antecedent in coclustering IV algorithm

Principales modifications dans
KWDataGridOptimizer::VNSOptimizeVarPartDataGrid
-> ajout du boolean bWithoutAntecedent qui permet de supprimer l'utilisation de la grille antecedente.
La grille antecedente et la grille optimisee different pas leur innerAttributes.
Si bWithouAntecedent=true, la grille optimisee courante est utilisee en entree de la procedure GNS de generation de grille voisine
Cela necessite de construire une grille "cadre" dont les innerAttributes sont compatibles avec ceux de la grille optimisee.
La grille voisine de cette grille optimisee est ensuite optimisee.
Cela necessite egalement que la grille initiale de reference pour l'optimisation possede des innerAttributes identiques a ceux de la grille voisine.

L'etape de surtokenisation des innerAttributes n'est pas activee dans cette version intermediaire.
bNewProto = false donc OptimizeVarPartDataGrid historique avec boucle sur tokenisation et pas PROTO_OptimizeVarPartDataGrid avec un niveau de tokenisation initial

On suit a plusieurs niveaux la difference de cout entre la grille optimisee courante et son export dont les innerAttributes seraient merges.
On verifie que cette difference est nulle lors des appes dans CCCoclusteringBuilder et OptimizeGranularizedDataGrid

Nouvelles methodes :
KWDataGridManager::ExportDataGridWithMergedInnerAttributes qui permet de generer une grille en modifiant ses innerAttributes par sa version fusionnee
KWDataGridManager::InitialiseVarPartAttributeWithNewSurtokenisedInnerAttributes
KWDataGridManager::InitialiseVarPartAtributeWithMergedInnerAttributes

Problemes restant a resoudre avec bWithoutAntecedent = true;
- Fuites memoire : provenant de la methode ExportDataGridWithMergedInnerAttributes ?
- echec des require de coherence des couts : ensure(fabs(dDataGridTotalCost - GetDataGridCosts()->ComputeDataGridMergerTotalCost(this)) < dEpsilon);
dans le cas d'une grille courante sans attribut informatif

marcboulle committed:
Memory bug resolution and VarPartsShared removal

Modification de la gestion memoire des attributes internes
- supression systematique des methodes de type Set|GeVarPartsShared
- ajout d'un compteur de reference dans KWDGInnerAttributes
- gestion des references depuis KWDGAttribute pour detruire automatiquement cet objet
  partage s'il n'est plus utilise
- cela resout le bug memoire, dans les cas avec ou sans antecedent

Mise a niveau des jeux de test LearningTest\TestCoclustering\y_CoclusteringIV_Standard
- impact dans CCCoclusteringReport::ReadJSONAttributePartition pour la relecture
  d'attribut internes avec "defaultGroupIndex"

marcboulle committed:
Cost ensure bug resolution for non-informative grid

Le probleme etait lie a l'utilisation de la methode CopyInformativeDataGrid,
qui supprime les dimension de la grille et ne permet plus d'avoir le vrai cout du modele null
- KWDataGridMerger::OptimizeMerge: remplacement par CopyDataGrid
- KWDataGridOptimizer: supression de l'attribut inutile bCleanNonInformativeVariables
- impact sur les references de jeux de test, qui ont change

Reste un bug potentiel avec un ensure (commente) dans KWDataGridOptimizer::VNSOptimizeVarPartDataGrid

Correction d'un bug, detecte en mode debug
- Cf. LearningTest\TestCoclustering\Standard\Iris
- Correction dans CCCoclusteringBuilder::ComputePartHierarchies (initialisation de la rootPart dans le cas d'un attribut n'ayant qu'une seule partie)

- Nettoyage des commentaires de mise au point
- Suppression de la methode CopyInformativeDataGrid

Bilan des tests sur LearningTest
- Cela semble des details, avec parfois :
  - Petite difference sur la trajectoire d'optimisation
  - Quelque fois des resultats moins bons, du a une trajectoire d'optimisation differente
    - confirme avec un test sur une variante de KWDataGridOptimizer::GenerateNeighbourSolution
      - code experimental qui genere une solution a partie d'une copie de la partie informative
        de la grille en cours plutot que de la grille en cours
      - cela produit des resultats parfois meilleurs, parfois moins bons
  - Quelques fois des differences de cout, y compris pour une meme solution
    - en fait, les ancien cout etaient parfois faux, notamment pour la separation preparation const et data cost
    - confirme apres investigation sur AIDSAtomCounts2D

Bilan :
- La nouvelle version semble correcte
- Les references des jeux de donnees devront evoluer

Detail du bilan des tests :
TestKhiops	Standard	BenchIris			errors	9		Problem file types: .xls -> pas de difference de performance mais difference dans le nombre de UsedAttributes (attributs non informatifs)
TestKhiops	Standard	SpliceJunction2D	warnings	8	errors	396		Problem file types: .khj, .xls	-> variation de Level dans certaines paires de variables ce qui change l'ordre du tableau
TestKhiops	MissingValues	MissingValuesDataGrid			errors	1758		Problem file types: .kdic, .khj, .xls		-> le modele de grille embarque les variables non informatives : 5 variables supplementaires. 8 variables et non 12 du fait d'attributs non obligatoire mais le cout serait a revoir dans le cas d'une selection de variables
TestKhiops	SparseData	AIDSAtomCounts2D			errors	8		Problem file types: .khj, .xls	-> difference sur preparationCost et dataCost pour une paire de variable Categoriel, Categoriel
TestKhiops	ParallelTask	Preparation2DLimits	warnings	1						Portability: Recovery from varying patterns in user messages -> uniquement variation de memoire
TestKhiops	NewPriorV9	Poub-SideEffect	warnings	4	errors	62		Problem file types: .khj, .xls	-> difference dans les Level
TestKhiops	SmallInstability	AllReports			errors	4513		Problem file types: .kdic, .khj, .xls		-> pas d'impact performance, variation d'un des Level en classification generalisee / dans le .json distinguer le cas classification du cas classification memoriser une tache de pre-traitement
TestKhiops	SmallInstability	BugsNBCompression			errors	26390		Problem file types: .khj, .xls		-> level d'une paire inferieure
TestKhiops	SmallInstability	BugsRegressorInfEval	warnings	2	errors	29346		Problem file types: .khj, .xls		-> modification de level univarie dans le cas d'une regression
TestKhiops	SmallInstability	CPM2D			errors	1355		Problem file types: .khj, .xls	-> level d'une paire superieur
TestKhiops	SmallInstability	CensusTrainClassLimits	warnings	1	errors	8501		Problem file types: .khj, .xls		-> en regression, level d'une variable superieur
TestKhiops	SmallInstability	Preparation2DTicTacToe			errors	925		Problem file types: .khj, .xls		-> pas d'impact performance, un level de paire inferieur, un autre superieur
TestKhiops	SmallInstability	RegressionOrangeJuice	warnings	2	errors	29406		Problem file types: .kdic, .khj, .xls -> pas de modif de performance mais level de variables modifies
TestKhiops	Histograms	PredictionAgeMaxPartNumber			errors	28269		Problem file types: .khj, .xls		-> Level de pairs legere variation

TestCoclustering	Standard	Adult2varsTiny			errors	472		Problem file types: err.txt, .khc, .khcj -> coclustering distincts avec level final legerement inferieur
TestCoclustering	Standard	AdultMissing			errors	1		Problem file types: err.txt -> uniquement lie au non affichage d'un attribut non informatif
TestCoclustering	Standard	AdultSmall			errors	2		Problem file types: err.txt -> uniquement lie au non affichage d'un attribut non informatif
TestCoclustering	Standard	AllCCResultsApiMode			errors	10		Problem file types: err.txt -> parcours distincts mais meme optimum a la fin
TestCoclustering	Standard	AllCCResultsStandardMode			errors	10		Problem file types: err.txt -> parcours distincts mais meme optimum a la fin
TestCoclustering	Standard	Iris			errors	5		Problem file types: err.txt -> parcours distincts mais meme optimum a la fin
TestCoclustering	Standard	IrisMissing			errors	3		Problem file types: err.txt -> parcours distincts mais meme optimum a la fin
TestCoclustering	Bugs	FormatFichier			errors	16997		Problem file types: err.txt, .khc, .khcj -> coclustering distincts avec level final legerement inferieur
TestCoclustering	SmallInstability	HierarchicalRanking			errors	2838		Problem file types: err.txt, .khc, .khcj -> coclustering distincts avec level final legerement inferieur
TestCoclustering	SmallInstability	NumericalRounding	warnings	228	errors	844		Problem file types: err.txt, .khc, .khcj -> coclustering distincts avec level final legerement superieur

TestCoclustering	y_CoclusteringIV_Standard	AdultEducation			errors	795		Problem file types: err.txt, .khc, .khcj -> coclustering distincts avec level final legerement inferieur
TestCoclustering	y_CoclusteringIV_Standard	IrisMissingValues			errors	750		Problem file types: err.txt, .khc, .khcj -> 1 cas sur 4 : coclustering distincts avec level final legerement inferieur
TestCoclustering	y_CoclusteringIV_Standard	LED			errors	60		Problem file types: err.txt	-> parcours distincts mais meme optimum a la fin
TestCoclustering	y_CoclusteringIV_Standard	Pima			errors	3656		Problem file types: err.txt, .khc, .khcj -> coclustering distincts avec level final legerement inferieur
TestCoclustering	y_CoclusteringIV_Standard	Tictactoe	warnings	68	errors	3534		Problem file types: .khc, .khcj	-> meme level et nombre de partie mais json distincts
TestCoclustering	y_CoclusteringIV_Standard	TooLongAdult2vars			errors	26296		Problem file types: err.txt, .khc, .khcj -> coclustering distincts avec level final legerement inferieur

carinehue committed:
Update Khiops version to 10.5.6-b.0

carinehue committed:
Update CICD LearningTest referencees

Add LearningTest/TestCoclustering/IV_Standard in 'full' family of LearningTest
Add IV_Iris basic IV test to test/LearningTest/TestCoclustering/Standard

marcboulle reviewed (changes requested):
Quelques suggestions de modification des noms et prototypes pour les nouvelles méthodes, à discuter, puis à prendre en compte éventuellement dans un nouveau point de commit.

Sinon, éventuellement supprimer les warnings suivants de KWDataGridOptimizer::PROTOGenerateNeighbourSolution
- `Learning\KWDataPreparation\KWDataGridOptimizer.cpp(2123): warning C4101: 'nMaxAttributeNumber' : variable locale non référencée`
- `Learning\KWDataPreparation\KWDataGridOptimizer.cpp(2122): warning C4101: 'nMandatoryAttributeNumber' : variable locale non référencée`

Et commencer à préparer la prochaine issue pour la suite des travaux sur la suppression des antécédents.
Par exemple: `Finalize removing management of "antecedent" in coclustering IV`
avec ce qui reste à faire, plus nettoyage des balises `// CH IV...`, `// CH ...`

carinehue committed:
Improve code according to PR requirements

Manage inner attributes as const for better reliability
Restore KWDataGridMerger::KWDGMPartMergeCompare method with dev version
Some cleaning

marcboulle reviewed (approved):
LGTM
